pkg/oc/cli/cmd/registry/login/login.go

```diff
@@ -0,0 +1,306 @@
+package login
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+
+	"github.com/golang/glog"
+	"github.com/spf13/cobra"
+
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apirequest "k8s.io/apiserver/pkg/endpoints/request"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/util/homedir"
+	"k8s.io/kubernetes/pkg/kubectl/cmd/templates"
+	kcmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+
+	imageclient "github.com/openshift/client-go/image/clientset/versioned"
+	imageapi "github.com/openshift/origin/pkg/image/apis/image"
+	"github.com/openshift/origin/pkg/image/registryclient"
+	"github.com/openshift/origin/pkg/oc/cli/util/clientcmd"
+)
+
+var (
+	loginDesc = templates.LongDesc(`
+		Login to the OpenShift integrated registry.
+
+		This logs your local Docker client into the OpenShift integrated registry using the
+		external registry name (if configured by your administrator). You may also log in
+		using a service account if you have access to its credentials. If you are logged in
+		to the server using a client certificate the command will report an error because
+		Docker commands do not generally allow client certificates.
+		
+		You may specify an alternate file to write credentials to with --to instead of 
+		.docker/config.json in your home directory. If you pass --to=- the file will be 
+		written to standard output.
+
+		To detect the registry hostname the client will attempt to find an image stream in
+		the current namespace or the openshift namespace and use the status fields that
+		indicate the registry hostnames. If no image stream is found or if you do not have
+		permission to view image streams you will have to pass the --registry flag with the 
+		desired hostname.
+
+		Experimental: This command is under active development and may change without notice.`)
+
+	loginExample = templates.Examples(`
+# Log in to the integrated registry
+%[1]s
+
+# Log in as the default service account in the current namespace
+%[1]s -z default
+`)
+)
+
+type Credentials struct {
```

Comment:
Wouldn't it be better to use `k8s.io/kubernetes/pkg/credentialprovider.DockerConfigJson` instead? Although it contains only reading bits for dockerfile it's still relevant, imho.

Reply:
This is about holding credentials from interior code to output, but not actually serializing the data.  We direct modify JSON to avoid dropping fields.